### DISCUSSION TITLE
Add visionOS support to deliver

### DISCUSSION THREAD

**Discussion by JagCesar** [fastlane/fastlane#21802] [Ideas]
Would be cool if we could add `visionOS` support to deliver, currently we only support `ios`, `appletvos`, and `osx`.

**Comment by JagCesar** [comment]
Played around with `pilot` and would love to have `visionOS` support here too.

**Reply by rogerluan** [reply]
Support to _pilot_ has been added here https://github.com/fastlane/fastlane/pull/21841 thanks to @demonnic 🚀 It will be included in the next _fastlane_ release :)

**Reply by alpennec** [reply]
Hey @rogerluan,

Can you please see my comment here: https://github.com/fastlane/fastlane/pull/21841#issuecomment-2186501421
I'm not able to use `xros` as a value for the `platform` parameter using fastlane 2.221.1.

`Error setting value 'xros' for option 'platform'`
`[!] The platform can only be ios, appletvos, or osx`

Is there anything specific to do? Thanks!

Thanks,
Axel

**Reply by mesqueeb** [reply]
@rogerluan can we not yet upload metadata for visionOS app?

I can download metadata but not upload...
```
[21:32:13]: Uploading metadata to App Store Connect for localized version 'ja'
[21:32:16]: Uploading app review information to App Store Connect
[21:32:17]: Running precheck before submitting to review, if you'd like to disable this check you can set run_precheck_before_submit to false
[21:32:17]: Running precheck 👮‍♀️ 👮
[21:32:17]: Error setting value 'xros' for option 'platform'
+---------------------------------+
|          Lane Context           |
+---------------+-----------------+
| PLATFORM_NAME |                 |
| LANE_NAME     | letsgo_visionos |
+---------------+-----------------+
[21:32:17]: Called from Fastfile at line 35
[21:32:17]: ```
[21:32:17]:     33:
[21:32:17]:     34:	lane :letsgo_visionos do
[21:32:17]:  => 35:		upload_to_app_store(
[21:32:17]:     36:			app_identifier: "studio.craftingtable.LetsGo",
[21:32:17]:     37:			platform: "xros",
[21:32:17]: ```
[21:32:17]: The platform can only be ios, appletvos, or osx

+------------------------------------------+
|             fastlane summary             |
+------+---------------------+-------------+
| Step | Action              | Time (in s) |
+------+---------------------+-------------+
| 💥   | upload_to_app_store | 35          |
+------+---------------------+-------------+

[21:32:17]: fastlane finished with errors

[!] The platform can only be ios, appletvos, or osx
```

**Comment by triplef** [comment]
This was added in https://github.com/fastlane/fastlane/pull/21574 and should be part of the next release (hopefully soon?!). 🎉 

Edit: I guess the PR is mainly for gym, tho it also adds visionOS to `supported_platforms` globally – not sure if that’s enough for deliver?

**Reply by demonnic** [reply]
I think the error being generated when I try to upload a vision pro app comes from pilot/lib/pilot/manager.rb and I don't see that being edited by the PR in question. But I'm not sure if it just needs adding to the list of acceptable platforms or if additional lift is required beyond that yet. With the hardware releasing next week it would definitely be nice to have though.

**Reply by rogerluan** [reply]
Answering the question, I think the referred PR won't solve the issue with uploading visionOS binary to the App Store. Other changes are required for that, probably in many places that we happen to mention `appletvos`, for instance:

<img width="325" alt="image" src="https://github.com/fastlane/fastlane/assets/8419048/1475a3a4-440c-40fc-858e-51e97695bb91">

That's my guess, but it needs testing, and updating the specs too. PRs are welcome, if anyone is feeling adventurous! ❤️

**Reply by demonnic** [reply]
I have made an attempt at a PR to add it to _pilot_ as that's what I actually use for my own uploading needs and could readily test. It did work for uploading a vision pro ipa though.

https://github.com/fastlane/fastlane/pull/21841

**Reply by rogerluan** [reply]
Thanks for this @demonnic ! 😊

**Reply by demonnic** [reply]
Happy to help! I'm willing to go in and do the same for deliver, but since I don't make use of it myself I'd need someone else to do the functional/practical testing. Was actually going to look at that later today.

**Reply by triplef** [reply]
Thank you @demonnic! I could probably test deliver.

**Reply by JagCesar** [reply]
I could also test, we currently use deliver to upload metadata and screenshots to a single app that runs natively on iOS, Apple Watch, tvOS, macOS. Would love to also upload all visionOS content using deliver 😄

**Reply by demonnic** [reply]
@triplef @JagCesar 
https://github.com/fastlane/fastlane/pull/21864
Hot off the presses. Honestly just went in and added xros with a crowbar everywhere appletvos appeared. So would definitely appreciate some testing. Hope this helps.

**Reply by rogerluan** [reply]
Go team!! 🔥

**Reply by JagCesar** [reply]
I tried my best running your branch @demonnic, but I just can't figure it out. I followed the instructions over at `https://github.com/fastlane/fastlane/blob/master/Testing.md#test-your-local-fastlane-code-base-with-your-setup`, but I still get the following error:

```
bundler: failed to load command: fastlane (/Users/cesar/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/bin/fastlane)
/Users/cesar/Developer/fastlane/spaceship/lib/spaceship/connect_api.rb:105:in `map': \e[31m[!] Cannot find a matching platform for 'xros' - valid values are IOS, MAC_OS, TV_OS, WATCH_OS\e[0m (RuntimeError)
```

**Reply by demonnic** [reply]
I'll take a look, was searching around for appletvos so could very well have missed some spots. I'm heading out for a bit but will dig in tonight

**Reply by alpennec** [reply]
I want to use Deliver to download/upload metadata + screenshots to App Store Connect for a visionOS app.

Curious to know if there are updates on this issue? how can we help? Thanks.
